frontend/src/components/editor/output/MarimoErrorOutput.tsx

```diff
@@ -68,193 +51,356 @@ export const MarimoErrorOutput = ({
+      e.type === "ancestor-stopped",
+  );
+  const syntaxErrors = errors.filter(
+    (e): e is Extract<MarimoError, { type: "syntax" }> => e.type === "syntax",
+  );
+  const unknownErrors = errors.filter(
+    (e): e is Extract<MarimoError, { type: "unknown" }> => e.type === "unknown",
+  );
+
+  const renderMessages = () => {
+    const messages: JSX.Element[] = [];
+
+    if (syntaxErrors.length > 0 || unknownErrors.length > 0) {
+      messages.push(
+        <li key="syntax-unknown">
+          {syntaxErrors.map((error, idx) => (
+            <p key={`syntax-${idx}`}>{error.msg}</p>
+          ))}
+          {unknownErrors.map((error, idx) => (
+            <p key={`unknown-${idx}`}>{error.msg}</p>
+          ))}
+          {cellId && (
+            <AutoFixButton
+              errors={[...syntaxErrors, ...unknownErrors]}
+              cellId={cellId}
+            />
+          )}
+        </li>,
+      );
+    }
+
+    if (cycleErrors.length > 0) {
+      messages.push(
+        <li key="cycle">
+          <p className="text-muted-foreground font-medium">
+            This cell is in a cycle.
+          </p>
+          <ul className="list-disc">
+            {cycleErrors.flatMap((error, errorIdx) =>
+              error.edges_with_vars.map((edge, edgeIdx) => (
+                <li className={liStyle} key={`cycle-${errorIdx}-${edgeIdx}`}>
                   <CellLinkError cellId={edge[0] as CellId} />
                   <span className="text-muted-foreground">
                     {" -> "}
-                    {edge[1].length === 1 ? edge[1] : edge[1].join(", ")}
+                    {edge[1].length === 1 ? edge[1][0] : edge[1].join(", ")}
                     {" -> "}
                   </span>
                   <CellLinkError cellId={edge[2] as CellId} />
                 </li>
-              ))}
-            </ul>
-            <Tip>Merge these cells into a single cell.</Tip>
-          </Fragment>
-        );
-
-      case "multiple-defs":
-        return (
-          <Fragment key={idx}>
-            <p className="text-muted-foreground">{`The variable '${error.name}' was defined by another cell:`}</p>
-            <ul className="list-disc">
-              {error.cells.map((cid) => (
-                <li className={liStyle} key={cid}>
-                  <CellLinkError cellId={cid as CellId} />
-                </li>
-              ))}
-            </ul>
-            <Tip>
-              <p className="pb-2">
-                marimo requires that each variable is defined in just one cell.
-                This constraint enables reactive and reproducible execution,
-                arbitrary cell reordering, seamless UI elements, execution as a
-                script, and more.
-              </p>
+              )),
+            )}
+          </ul>
+          {cellId && <AutoFixButton errors={cycleErrors} cellId={cellId} />}
+          <Tip title="Resolving cycles" className="mb-2">
```

Comment:
I think this copy could do with an update too?
Ideas:

 - Why does my code have cycles, and how can I resolve them?
 - How do I remove cycles?
 - Why can't I have cycles?

Phrasing the help as a question works in the latter case

Reply:
Thanks, done!